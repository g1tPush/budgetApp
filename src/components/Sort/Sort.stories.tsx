import React from 'react'
import { ComponentStory, ComponentMeta } from '@storybook/react'
import { useState } from 'react'

import Sort from './Sort'

export default {
    title: 'Sort Header',
    component: Sort
} as ComponentMeta<typeof Sort>

export const SortElement: ComponentStory<typeof Sort> = (args) => <Sort />
