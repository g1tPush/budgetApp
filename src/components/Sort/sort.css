@media only screen and (max-width: 750px) {
    .DateInput {
        width: 100px !important;
    }
}

@media only screen and (max-width: 350px) {
    .DateRangePicker_picker__directionLeft {
        left: -19px !important;
    }
}