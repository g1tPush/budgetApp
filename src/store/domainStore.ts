import { makeObservable, observable, makeAutoObservable } from 'mobx'
import moment from 'moment'
import { IExpense, SortType, DateType } from '../types/types'

// {
//     id: '123ew7t23ieuwd',
//     content: 'Iphone',
//     amount: 1000,
//     date: moment(),
//     note: 'This is a good choise'
// }

export class ExpensesStore {
    expenses: Expense[] = []
    searchText: string = ''
    type: SortType = 'Date'
    startDate: DateType = null
    endDate: DateType = null
    filterredArr: Expense[] = []

    constructor() {
        makeObservable(this, {
            expenses: observable,
            searchText: observable,
            type: observable,
            startDate: observable,
            endDate: observable,
            filterredArr: observable
        })
    }

    get expensesList() {
        return this.expenses.slice().filter((item) => {
            if (item.date) {
                const date = moment(item.date).format('MM-DD-YYYY')
                if (this.startDate === null && this.endDate === null) {
                    return item
                } else if (this.startDate !== null && this.endDate !== null) {
                    return new Date(date) <= new Date(this.endDate.format('MM-DD-YYYY')) && new Date(date) >= new Date(this.startDate.format('MM-DD-YYYY'))
                } else if (this.startDate === null) {
                    return new Date(date) <= new Date(this.endDate?.format('MM-DD-YYYY') || '')
                } else if (this.endDate === null) {
                    return new Date(date) >= new Date(this.startDate?.format('MM-DD-YYYY') || '')
                }
            }
            return 1
        }).filter((item) => {
            return item.content.toLowerCase().includes(this.searchText.toLowerCase())
        }).sort((a, b) => {
            if (b.date && a.date && this.type === 'Date') {
                return b.date?.valueOf() - a.date?.valueOf()
            } else if (this.type === 'Amount') {
                return +(b.amount) - +(a.amount)
            }
            return 1
        })
    }

    updateDate(startDate: DateType, endDate: DateType) {
        this.startDate = startDate
        this.endDate = endDate
    }

    updateType(type: 'Date' | 'Amount') {
        this.type = type
    }

    updateSearchText(text: string) {
        this.searchText = text
    }

    createExpense(expense: IExpense) {
        this.expenses.push(new Expense(expense))
        return expense
    }

    updateExpense(expense: IExpense) {
        let index = this.expenses.findIndex(el => {
            return el.id === expense.id
        })

        this.expenses[index] = expense
    }

    removeExpense(expense: Expense | undefined) {
        if (expense) {
            const index = this.expenses.findIndex((el) => {
                return el.id === expense.id
            })
            this.expenses.splice(index, 1)
        }
    }

    getExpense(expenseId: string | undefined) {
        if (expenseId === undefined) {
            return
        }
        return this.expenses.slice().find(el => {
            return el.id === expenseId
        })
    }
}

export class Expense {
    id: string | null = null
    content: string = ''
    date: DateType = null
    note: string;
    amount: number = 0

    constructor(store: IExpense) {
        makeAutoObservable(this, {
            id: false,
            content: false,
            date: false,
            amount: false
        })


        this.id = store.id
        this.amount = store.amount
        this.content = store.content
        this.note = store.note
        this.date = store.date
    }
}