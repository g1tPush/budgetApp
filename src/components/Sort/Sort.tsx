/* eslint-disable react-hooks/exhaustive-deps */
import React from 'react'
import { useEffect } from 'react'
import 'react-dates/initialize'
import { DateRangePicker, FocusedInputShape } from 'react-dates'
import 'react-dates/lib/css/_datepicker.css'
import Input from '../Input/Input'
import styled from 'styled-components'
import { useStore } from '../../index'
import { observer } from 'mobx-react-lite'
import './sort.css'

const Wrap = styled.div`
    display: flex;
    gap: 15px;
    @media (max-width: 750px) {
        gap: 10px;
        flex-flow: row wrap;
    }
`

const Select = styled.select`
    height: 50px;
    width: 100px;
    background-color: white;
`

const Sort = () => {
    const expenses = useStore()

    const [text, setText] = React.useState('')
    const startDate = expenses.startDate
    const endDate = expenses.endDate
    const [focusedInput, setFocusedInput] = React.useState<FocusedInputShape | null>(null)

    const update = (e: React.ChangeEvent<HTMLInputElement>) => {
        setText(e.target.value)
    }

    useEffect(() => {
        const timer = setTimeout(() => {
            expenses.updateSearchText(text)
        }, 750)

        return () => {
            clearTimeout(timer)
        }
    }, [text])

    return (
        <Wrap>
            <Input value={text} type='text' placeholder='Search Expenses' onChange={update} />
            <Select value={expenses.type} onChange={(e) => {
                expenses.updateType(e.target.value === 'Date' ? 'Date' : 'Amount')
            }}>
                <option value="Date">Date</option>
                <option value="Amount">Amount</option>
            </Select>
            <DateRangePicker
                startDate={startDate}
                isOutsideRange={() => false}
                startDateId='start-date'
                endDate={endDate}
                displayFormat={() => "DD-MM-YYYY"}
                endDateId='end-date'
                onDatesChange={async ({ startDate, endDate }) => {
                    await expenses.updateDate(startDate, endDate)
                }}
                focusedInput={focusedInput}
                onFocusChange={(focusedInput) => setFocusedInput(focusedInput)}
                showClearDates={true}
                numberOfMonths={1}
            />
        </Wrap>
    )
}

export default observer(Sort)