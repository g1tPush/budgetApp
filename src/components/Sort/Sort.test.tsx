import React from 'react'
import renderer from 'react-test-renderer'
import Sort from './Sort'
import { StoreProvider } from '../../index'
import { ExpensesStore } from '../../store/domainStore'

it('renders correctly', () => {
    const tree = renderer
        .create(
            <StoreProvider value={new ExpensesStore()}>
                <Sort />
            </StoreProvider>
        )
        .toJSON()
    expect(tree).toMatchSnapshot()
})