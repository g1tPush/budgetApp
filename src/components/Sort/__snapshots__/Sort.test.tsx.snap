// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="sc-hAZoDl fsXIpB"
>
  <input
    className="sc-bczRLJ favxIM"
    onChange={[Function]}
    placeholder="Search Expenses"
    type="text"
    value=""
  />
  <select
    className="sc-fnykZs bVozEv"
    onChange={[Function]}
    value="Date"
  >
    <option
      value="Date"
    >
      Date
    </option>
    <option
      value="Amount"
    >
      Amount
    </option>
  </select>
  <div
    className="DateRangePicker DateRangePicker_1"
  >
    <div>
      <div
        className="DateRangePickerInput DateRangePickerInput_1 DateRangePickerInput__withBorder DateRangePickerInput__withBorder_2 DateRangePickerInput__showClearDates DateRangePickerInput__showClearDates_3"
      >
        <div
          className="DateInput DateInput_1"
        >
          <input
            aria-describedby="DateInput__screen-reader-message-start-date"
            aria-label="Start Date"
            autoComplete="off"
            className="DateInput_input DateInput_input_1"
            disabled={false}
            id="start-date"
            name="start-date"
            onChange={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            placeholder="Start Date"
            readOnly={false}
            required={false}
            type="text"
            value=""
          />
          <p
            className="DateInput_screenReaderMessage DateInput_screenReaderMessage_1"
            id="DateInput__screen-reader-message-start-date"
          >
            Navigate forward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.
          </p>
        </div>
        <div
          aria-hidden="true"
          className="DateRangePickerInput_arrow DateRangePickerInput_arrow_1"
          role="presentation"
        >
          <svg
            className="DateRangePickerInput_arrow_svg DateRangePickerInput_arrow_svg_1"
            focusable="false"
            viewBox="0 0 1000 1000"
          >
            <path
              d="M694 242l249 250c12 11 12 21 1 32L694 773c-5 5-10 7-16 7s-11-2-16-7c-11-11-11-21 0-32l210-210H68c-13 0-23-10-23-23s10-23 23-23h806L662 275c-21-22 11-54 32-33z"
            />
          </svg>
        </div>
        <div
          className="DateInput DateInput_1"
        >
          <input
            aria-describedby="DateInput__screen-reader-message-end-date"
            aria-label="End Date"
            autoComplete="off"
            className="DateInput_input DateInput_input_1"
            disabled={false}
            id="end-date"
            name="end-date"
            onChange={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            placeholder="End Date"
            readOnly={false}
            required={false}
            type="text"
            value=""
          />
          <p
            className="DateInput_screenReaderMessage DateInput_screenReaderMessage_1"
            id="DateInput__screen-reader-message-end-date"
          >
            Navigate backward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.
          </p>
        </div>
        <button
          aria-label="Clear Dates"
          className="DateRangePickerInput_clearDates DateRangePickerInput_clearDates_1 DateRangePickerInput_clearDates_default DateRangePickerInput_clearDates_default_2 DateRangePickerInput_clearDates__hide DateRangePickerInput_clearDates__hide_3"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          <svg
            className="DateRangePickerInput_clearDates_svg DateRangePickerInput_clearDates_svg_1"
            focusable="false"
            viewBox="0 0 12 12"
          >
            <path
              d="M11.53.47a.75.75 0 0 0-1.061 0l-4.47 4.47L1.529.47A.75.75 0 1 0 .468 1.531l4.47 4.47-4.47 4.47a.75.75 0 1 0 1.061 1.061l4.47-4.47 4.47 4.47a.75.75 0 1 0 1.061-1.061l-4.47-4.47 4.47-4.47a.75.75 0 0 0 0-1.061z"
              fillRule="evenodd"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</div>
`;
